torch/utils/data/datapipes/_typing.py

```diff
@@ -390,24 +428,29 @@ def __next__(self):
             # TODO: Add try-except to in-place reduce traceback from the Exception
             # See: https://github.com/pytorch/data/issues/284
             with context():
+                _check_iterator_valid(self.source_dp, self.iterator_id)
                 return next(self.iterator)
 
         def __getattr__(self, name):
             return getattr(self.iterator, name)
 
     func = namespace['__iter__']
+    iterator_id: Optional[int] = None  # This ID is tied to each created iterator
 
     # ``__iter__`` of IterDataPipe is a generator function
     if inspect.isgeneratorfunction(func):
         @functools.wraps(func)
         def wrap_generator(*args, **kwargs):
             gen = func(*args, **kwargs)
+            datapipe = args[0]
+            iterator_id = _set_datapipe_valid_iterator_id(datapipe)
```

Comment:
Could we make sure `reset` will be invoked everytime `iter` is called?

Reply:
One idea is to check if the DataPipe `hasattr(dp, "reset")` and call it here if it does. I am open to do that instead of relying on `__iter__` to call `reset` (which is what a PR later in the stack is doing).

Reply:
I prefer to add a noop `reset` to base class and always call it during `__iter__` to save users' effort. And, adding a contribution doc to indicate `reset` is required.

Reply:
Okay, let me try and add that to the PR later in the stack.